### ISSUE TITLE
Express SQL projection by specifying only columns to exclude

### ISSUE THREAD

**Issue by healiseu** [ClickHouse/ClickHouse#5109]
ClickHouse is a true column-oriented DBMS therefore I thought it makes sense to be able to express projection of many columns using an alternative syntax. Instead of specifying a long comma list of columns excluding those you don't wish to include in the projection:
```sql
SELECT col1, col2, col3, col4, col5, col6, col7, ... col999 FROM test
```
You could use an operator such as ALL BUT (see TutorialD, [reldb][1] and [Dataphor][2] bnf syntax) with a short list of columns to exclude from projection e.g.
```sql
SELECT ALL BUT col452, col611, col56 FROM test
```
It seems developers of existing versions of SQL language are reluctant to include such a feature in their DBMS, and I guess one basic reason is that these are NOT columnar DBMS. Although there are tricks out there to do this in one way or another ([stackoverflow1][3], [stackoverflow2][4], [stackoverflow3][5]) these are dirty tricks and not elegant or handy at all. For example, create a temporary table copy the data and then use ALTER DROP for those columns you wish to exclude from your data set.

So, what do you think ? What are the obstacles to implement such a feature in ClickHouse SQL dialect ?

As it concerns its use, I think it's quite common in columnar DBMS to handle tables with tens or hundreds of columns so that can be a handy, nice and cool feature.

[1]: https://github.com/henriavelabarbe/reldb
[2]: https://github.com/DBCG/Dataphor/blob/master/Documentation/DevelopersGuide/D4TableOperators.adoc
[3]: https://stackoverflow.com/questions/3650689/sql-exclude-a-column-from-select-query
[4]: https://stackoverflow.com/questions/729197/sql-exclude-a-column-using-select-except-columna-from-tablea
[5]: https://stackoverflow.com/questions/29095281/how-to-select-all-the-columns-of-a-table-except-one-column/29114381

**Comment by geldot** [comment]
+1. Additionally, having `SELECT ALL` (with or without the `BUT` clause) include virtual columns would be a welcome feature for columnar workflows.

**Comment by geldot** [comment]
The COLUMNS() matcher from #6038 is merged. However it can only select, not deselect, columns due to re2 regex syntax limitations. So a way to negate is still needed. COLUMNS_EXCEPT()?

**Comment by alexey-milovidov** [comment]
After columns matcher was introduces, this task looks like an extension that is easy to implement.

**Comment by alexey-milovidov** [comment]
The feature has been implemented by @amosbird in #14233

**Comment by sergzach** [comment]
> After columns matcher was introduced, this task looks like an extension that is easy to implement.

It's very unclear because `(?!re)` is not supported.
